### PR TITLE
Assert property types nested

### PR DESCRIPTION
**Pull request by yarneo**
assert property types for nested types, upto 1000 elements long and 3 levels of nesting deep

### CODE REVIEW

File: Mixpanel/Mixpanel.m [hunk 1]
```diff
@@ -181,20 +181,57 @@ - (void)setUseIPAddressForGeoLocation:(BOOL)useIPAddressForGeoLocation {
...
+        [Mixpanel assertPropertyTypesInDictionary:propertyValue depth:depth+1];
+    } else if ([propertyValue isKindOfClass:[NSArray class]]) {
+        [Mixpanel assertPropertyTypesInArray:propertyValue depth:depth+1];
+    }
+}
+
++ (void)assertPropertyTypesInDictionary:(NSDictionary *)properties depth:(NSUInteger)depth
+{
+    if([properties count] > 1000) {
+        MPLogWarning(@"You have an NSArray in your properties that is bigger than 1000 in size. \
```
Comment (pchien) [inline Mixpanel/Mixpanel.m]: doesn't matter much but it says NSArray here

Comment (yarneo) [inline Mixpanel/Mixpanel.m]: good catch

**Comment by LorienMan** [comment]
Great improvement! But I doubt that such constraints like `count` or `depth` really necessary because a single property of wrong type can lead to cycle-crash on Mixpanel start and we'd like to catch this problem easily. As `NSAssert`s work only in debug builds it is not so bad to perform full recursive check (without constraints) only in debug builds where performance impact would not be so harmful (but it will definitely help us find a problem).

**Comment by yarneo** [comment]
@LorienMan thanks for bringing this to our attention. We knew of the possibility of nested types to be faulty for a while, but refrained to do nested checks due to degrading performance by passing over every element. Therefore we put the depth and count as a limiting factor but as you can see they are also very large. We strongly recommend against sending arrays with more than 1,000 cells or very nested data structures. Hopefully in your scenario and in the other 99% of use-cases this will be a good enough check. If someone does choose to send very very large data structures as properties, we prefer they be responsible for the type checking rather than us degrading their app's performance. (Just to answer your side-note, yes NSAssert is only in debug builds but the iteration over every element to check its type is done in release builds as well). Thanks again for sending out the PR, we really appreciate the help!

**Comment by LorienMan** [comment]
@yarneo It is possible to make something like this:

```
+ (BOOL)assertPropertyTypesInDictionary:(NSDictionary *)properties;
+ (BOOL)assertPropertyTypesInArray:(NSArray *)array;

+ (void)assertPropertyType:(id)propertyValue
{
    // ......
    if ([propertyValue isKindOfClass:[NSDictionary class]]) {
        NSAssert([Mixpanel assertPropertyTypesInDictionary:propertyValue], nil);
    } else if ([propertyValue isKindOfClass:[NSArray class]]) {
        NSAssert([Mixpanel assertPropertyTypesInArray:propertyValue], nil);
    }
}

```

In release builds there will be no nested checks at all. In debug builds we will check everything to be of correct type.

**Comment by yarneo** [comment]
I love it! Gonna send an update to the PR :)

**Comment by pchien** [comment]
lgtm

**Pull request merged**